docs/syntax-if-statements-initializers.md

```diff
@@ -0,0 +1,114 @@
+
+The reduced scope pollution improves register space in extreme cases (or auto-generated code) where developers have many variables defined and have to work around register limits by reducing the register size. In some scenarios, especially on Roblox, initializing variables within if statements can lead to improved performance and reduced complexity by avoiding unnecessary calls by developers. ⁤ A common paradigm used by Roblox developers is to use `Instance:FindFirstChild` in their condition and then access it afterwards instead of keeping an existing variable around in the new scope, polluting the existing scope.
+
+# Design
+
+If statements with initializers must match (following the Luau grammar) `'if' 'local' bindinglist ['=' explist] 'then'` and `'local' bindinglist ['=' explist] where exp 'then'` syntax. The variables declared by the initializer are only available to the block of that condition and will be undefined to the `elseif` conditions and blocks as well as the `else` block; any code after the if statement won't have the variables defined either.
+
+In the former case, the value of the first declared variable will be checked.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() then
+   print(b)
+else
+   print("false")
+end
+```
+
+`Output: true`
+
+In the latter case, the `exp` condition is checked rather than the initializer.
+
+Example:
+
+```lua
+local function foo()
+    return true
+end
+
+if local b = foo() where b == false then
+   print(b)
+else
+   print("true")
+end
+```
+
+`Output: true`
+
+When declaring multiple values inside of a condition, all of the variables will be evaluated.
+
+Example:
+
+```lua
+local function foo()
+    return true, false
+end
+
+if local a,b = foo() then
+else
+    print'Hello World, from Luau!'
+end
+```
+
+`Output: Hello World, from Luau!`
+
+If statement initializers are also allowed in `elseif` conditions.
+
+Example:
+
+```lua
+local a = false
+local function foo()
+    local b = a
+    a = true
+    return b
+end
+
+if local a = foo() then
+elseif local b = foo() then
+    print(b)
+end
+```
+
+`Output: true`
+
+# Drawbacks
+
+Parser recovery may be more fragile due to the `local` keyword.
+
+Initializers increase the complexity of the language syntax and may obscure control flow in complex conditions.
+
+# Alternatives
+
+A different keyword or token can be used in place of `where`.
+
+Introducing a new contextual keyword can introduce complexity in the parser, and a different existing keyword like `in` or `do` could be used instead.
```

Comment:
As a joke, I also thought of using `if` twice, e.g. `if local a, b = foo() if a > b then`. It's not a good idea though in practice.